src/client/browser/idle-page/index.js

```diff
@@ -3,36 +3,46 @@ import StatusIndicator from './status-indicator';
 import COMMAND from '../../../browser/connection/command';
 
 
-const CHECK_STATUS_DELAY = 1000;
-
 const createXHR = () => new XMLHttpRequest();
 
+const CHECK_STATUS_DELAY = 1000;
 
 class IdlePage {
-    constructor (statusUrl, heartbeatUrl, initScriptUrl, options = {}) {
+    constructor (statusUrl, heartbeatUrl, initScriptUrl, { retryTestPages } = {}) {
         this.statusUrl       = statusUrl;
+        this.heartbeatUrl    = heartbeatUrl;
+        this.initScriptUrl   = initScriptUrl;
         this.statusIndicator = new StatusIndicator();
 
-        if (options.retryTestPages)
-            browser.enableRetryingTestPages();
+        this.retryTestPages = retryTestPages;
 
-        browser.startHeartbeat(heartbeatUrl, createXHR);
-        browser.startInitScriptExecution(initScriptUrl, createXHR);
+        document.title = '[' + document.location.toString() + ']';
+    }
 
-        this._checkStatus();
+    async _pollStatus () {
+        let { command } = await browser.checkStatus(this.statusUrl, createXHR);
 
-        document.title = '[' + document.location.toString() + ']';
+        while (command.cmd === COMMAND.idle) {
+            await browser.delay(CHECK_STATUS_DELAY);
+
+            ({ command } = await browser.checkStatus(this.statusUrl, createXHR));
+        }
     }
 
-    async _checkStatus () {
-        try {
-            const { command } = await browser.checkStatus(this.statusUrl, createXHR);
+    async start () {
+        if (this.retryTestPages)
+            await browser.enableRetryingTestPages();
 
-            if (command.cmd === COMMAND.idle)
-                window.setTimeout(() => this._checkStatus(), CHECK_STATUS_DELAY);
+        browser.startHeartbeat(this.heartbeatUrl, createXHR);
+        browser.startInitScriptExecution(this.initScriptUrl, createXHR);
+
+        try {
+            await this._pollStatus();
         }
-        catch (err) {
+        catch (error) {
             this.statusIndicator.showDisconnection();
+
+            throw error;
```

Comment:
Which subsystem will handle this error?

Reply:
Hmm... As I see in the previous version this error was caught and not rethrown. As far as I understand now we rethrow explicitly. @AndreyBelym could you add more details on this?

Reply:
There is not so much point about it. While debugging, I thought that it would be useful to see an error in the browser console when the browser finally gives up and disconnects. Except for an error in the console, the result is the same as in the previous version - when the browser disconnects, we lost him until someone somehow connects it back manually vi reloading or whatever.